lib/evmone/tracing.cpp

```diff
@@ -64,10 +66,103 @@ class HistogramTracer : public Tracer
 public:
     explicit HistogramTracer(std::ostream& out) noexcept : m_out{out} {}
 };
+
+
+class InstructionTracer : public Tracer
+{
+    struct Context
+    {
+        const uint8_t* const code;  ///< Reference to the code being executed.
+        const int64_t start_gas;
+        const char* const* const opcode_names;
+
+        Context(const uint8_t* c, int64_t g, const char* const* n) noexcept
+          : code{c}, start_gas{g}, opcode_names{n}
+        {}
+    };
+
+    std::stack<Context> m_contexts;
+    std::ostream& m_out;  ///< Output stream.
+
+    void output_stack(const Stack& stack)
+    {
+        const auto size = stack.size();
+        m_out << R"(,"stack":[)";
+        for (int i = 0; i < size; ++i)
+        {
+            if (i != 0)
+                m_out << ',';
+            m_out << R"("0x)" << to_string(stack[i], 16) << '"';
+        }
+        m_out << ']';
+    }
+
+    void on_execution_start(
+        evmc_revision rev, const evmc_message& msg, bytes_view code) noexcept override
+    {
+        m_contexts.emplace(code.data(), msg.gas, evmc_get_instruction_names_table(rev));
+
+        m_out << "{";
+        m_out << R"("kind":")"
+              << ((msg.kind == EVMC_CREATE || msg.kind == EVMC_CREATE2) ? "create" : "call") << '"';
+        m_out << R"(,"static":)" << (((msg.flags & EVMC_STATIC) != 0) ? "true" : "false");
+        m_out << R"(,"depth":)" << msg.depth;
+        m_out << R"(,"rev":")" << rev << '"';
+        m_out << "}\n";
+    }
+
+    void on_instruction_start(uint32_t pc, const ExecutionState& state) noexcept override
+    {
+        const auto& ctx = m_contexts.top();
+
+        const auto opcode = ctx.code[pc];
+        m_out << "{";
+        m_out << R"("pc":)" << pc;
+        m_out << R"(,"op":)" << int{opcode};
+        m_out << R"(,"opName":")" << get_name(ctx.opcode_names, opcode) << '"';
+        m_out << R"(,"gas":)" << state.gas_left;
+        output_stack(state.stack);
+
+        // Full memory can be dumped as evmc::hex({state.memory.data(), state.memory.size()}),
+        // but this should not be done by default. Adding --tracing=+memory option would be nice.
+        m_out << R"(,"memorySize":)" << state.memory.size();
+
+        m_out << "}\n";
+    }
+
+    void on_execution_end(const evmc_result& result) noexcept override
+    {
+        const auto& ctx = m_contexts.top();
+
+        m_out << "{";
+        m_out << R"("error":)";
+        if (result.status_code == EVMC_SUCCESS)
+            m_out << "null";
+        else
+            m_out << '"' << result.status_code << '"';
```

Comment:
I'm confused about where is this status code converted to string.

Reply:
The `evmc.hpp` defines `operator<<` for these. Can be changed to `to_string()`.